### PR TITLE
Add SAP and HA dashboards to grafana formula

### PR DESCRIPTION
**Pull request by MalloZup**
# desc:
This pr add sap and ha cluster dashboards maintained by SAP/ha team

I have followed the same pattern you have on the grafana-formula
note that I didn't tested this.

# Consideration/Request for comments:

Since we package already our dashboards, putting the json files here would be somehow not the cleanest solution since we would need to maintain `json` here AND in the source code of the original dashboards. ( so more then 2 places)

On the other side, I don't remember if the SUSE-Manager server has access to the SLES repository  for installing the RPM packages but I guess yes and that this approach should be ok.

let me know

### CODE REVIEW

File: grafana-formula/grafana/init.sls
Comment (stefanotorresi) [inline grafana-formula/grafana/init.sls:110]: I think the netweaver dashboard should go in a separate setting, we don't usually install them both in the same node.

Comment (MalloZup) [reply grafana-formula/grafana/init.sls:110]: ok fair point. thx

**Comment by MalloZup** [comment]
@renner @cavalheiro @witekest  what is your pov on this?

ju SLE15 submissions are done for the PKGs. can suma-server install such server or how can we install the dash? tia

File: grafana-formula/metadata/form.yml
Comment (renner) [inline grafana-formula/metadata/form.yml:94]: Maybe for consistency make it singular like the other states: `add_ha_dashboard`?

File: grafana-formula/metadata/form.yml
Comment (renner) [inline grafana-formula/metadata/form.yml:100]: I would suggest to follow @stefanotorresi recommendation from above and split it up in two states? Then I would suggest to write both as singular as well, e.g. `add_sap_hana_dashboard` and `add_sap_netweaver_dashboard`.

Comment (stefanotorresi) [reply grafana-formula/metadata/form.yml:100]: @renner the packages use the plural `dashboards` deliberately because we might provide multiple dashboards within one package.

Comment (renner) [reply grafana-formula/metadata/form.yml:100]: That's fine, feel free to keep the plural here as well in case it makes sense, i.e. more than one dashboard is added with a state.

Comment (MalloZup) [reply grafana-formula/metadata/form.yml:100]: yes is something on my todo list will push a commit on this

File: grafana-formula/metadata/form.yml
Comment (MalloZup) [reply grafana-formula/metadata/form.yml:94]: yes , originally were 3 make sense

**Comment by MalloZup** [comment]
@juliogonzalez cc

**Comment by juliogonzalez** [comment]
> On the other side, I don't remember if the SUSE-Manager server has access to the SLES repository for installing the RPM packages but I guess yes and that this approach should be ok.

It all the depends, on what channel(s) is this package?

**Comment by MalloZup** [comment]
@juliogonzalez it is on SLES codebase. Update sent to you on rocketchat

**Comment by juliogonzalez** [comment]
Uh-oh, I am afraid we are going to have a problem here.

SUSE manager its own `grafana-formula` on the SUSE Manager codestream: https://build.suse.de/package/show/SUSE:SLE-15-SP2:Update:Products:Manager41/grafana-formula

We can of course take it from SLES codestream as we are lucky and the version there is higher. But then @witekest needs to be aware that the maintenance will happen *there* by branching the package. And before doing that, I'd like to have his OK in case there's something I don't have in mind right now.

Do we *really* need this for 4.1.1 or can it wait for 4.1.2? I'd really like to avoid yet another last-minute-change, as this was not even at https://github.com/SUSE/spacewalk/issues/11942 or https://jira.suse.com/browse/ECO-2237 on the list of packages part of the ECO.

**Comment by renner** [comment]
> Uh-oh, I am afraid we are going to have a problem here.
> 
> SUSE manager its own `grafana-formula` on the SUSE Manager codestream: https://build.suse.de/package/show/SUSE:SLE-15-SP2:Update:Products:Manager41/grafana-formula
> 
> We can of course take it from SLES codestream as we are lucky and the version there is higher. But then @witekest needs to be aware that the maintenance will happen _there_ by branching the package. And before doing that, I'd like to have his OK in case there's something I don't have in mind right now.

As I understood, Dario was referring to the packages that would be installed with the `grafana-formula` after this patch (rather than `grafana-formula` itself):

- `grafana-ha-cluster-dashboards`
- `grafana-sap-hana-dashboards`
- `grafana-sap-netweaver-dashboards`

He submitted those to the SLES 15 channels, so in the current setup this part of the formula works only for systems that have those channels assigned. As mentioned before, I think a more consistent approach would be to either ship those dashboards as json files like the other ones, or provide all dashboards as packages with the same channels as we are providing Grafana itself, i.e. the client tools channels for SLE15 and SLE12.

> Do we _really_ need this for 4.1.1 or can it wait for 4.1.2? I'd really like to avoid yet another last-minute-change, as this was not even at [SUSE/spacewalk#11942](https://github.com/SUSE/spacewalk/issues/11942) or https://jira.suse.com/browse/ECO-2237 on the list of packages part of the ECO.

I don't think it's required for 4.1.1, we should probably aim for releasing this with 4.1.2 and try to get the mentioned packaging / linking done in order to support all platforms where we are support the provisioning of Grafana with this formula?

**Comment by juliogonzalez** [comment]
> I don't think it's required for 4.1.1, we should probably aim for releasing this with 4.1.2 and try to get the mentioned packaging / linking done in order to support all platforms where we are support the provisioning of Grafana with this formula?

I guess the most sensible solution is submitting this update of grafana-formula to the SLE15 codestreams, and having SUSE Manager taking it from there with 4.1.2 :-)

**Comment by renner** [comment]
> > I don't think it's required for 4.1.1, we should probably aim for releasing this with 4.1.2 and try to get the mentioned packaging / linking done in order to support all platforms where we are support the provisioning of Grafana with this formula?
> 
> I guess the most sensible solution is submitting this update of grafana-formula to the SLE15 codestreams, and having SUSE Manager taking it from there with 4.1.2 (via product definition for devel, or via channel definitions for the release) :-)

Julio, `grafana-formula` is a package that we are shipping with the SUSE Manager **Server** channels to make it appear in the UI. This should not change, we would just update the package with this patch for 4.1.2. The formula contains states for provisioning Grafana itself and dashboards on managed systems (that can be either SLE12 or SLE15). Therefore our client tools contain the `grafana` package and I guess we should be shipping those new packaged dashboards as well with those client tools? So we would need to somehow bring those packages that Dario submitted to SLE15 to our SLE12 and 15 client tools.

**Comment by juliogonzalez** [comment]
> Julio, grafana-formula is a package that we are shipping with the SUSE Manager Server channels to make it appear in the UI. This should not change, we would just update the package with this patch for 4.1.2. The formula contains states for provisioning Grafana itself and dashboards on managed systems (that can be either SLE12 or SLE15). Therefore our client tools contain the grafana package and I guess we should be shipping those new packaged dashboards as well with those client tools? So we would need to somehow bring those packages that Dario submitted to SLE15 to our SLE12 and 15 client tools.

Let me clarify. We have `grafana` in our client tools project (devel) because we maintain grafana. But we submit it to the SLE12 and SLE15 codestreams.

Now the dashboards are a different story, as they it seems they are not maintained by us, IIUC. Therefore we should add them to the SLE15 client tools product definitions and SLE15 client tools channel definitions, but there's no need to to link them or copypac them to our devel projects.

Now the question is what happens about other OS: Who is going to maintain the packages for those OS? Us or dario/sap? And where are we going to maintain them? At the OS codestreams so other projects can use them, or at the client tools codestreams (only available as separate codestreams for RES8 and Ubuntu20?

**Comment by MalloZup** [comment]
@juliogonzalez  so we are going to maintain the grafana-dashboards rpms packages.

However we never tested for RES8/Ubuntu, is this strictly mandatory or the grafana-formula can be more "elastic" in that sense?

As stated from @renner ,

I could submit the `json` files but this is a PITA for maintainers. Each time we change a dashboard which we develop we would need to submit a PR here. And imagine if Suse-manager would have like multiple dashboard from every team, this  would make lot of PRs.

If it is strictly mandatory to have RHEL/DEBian packages, we can help you to have such packages.


As you can see,  we have the RPM here which is automated by github-actions.
https://github.com/ClusterLabs/ha_cluster_exporter/blob/master/packaging/obs/grafana-ha-cluster-dashboards/grafana-ha-cluster-dashboards.spec#L16


The dashboard  package itself is just a json file

**Comment by renner** [comment]
> Now the question is what happens about other OS: Who is going to maintain the packages for those OS? Us or dario/sap? And where are we going to maintain them? At the OS codestreams so other projects can use them, or at the client tools codestreams (only available as separate codestreams for RES8 and Ubuntu20?

We are not actually supporting grafana itself for other OSs (RES or Ubuntu), only for SLE12 and SLE15. Therefore - from my perspective - we should just ship Dario's dashboard packages alongside the grafana package with our [SLE12](https://build.suse.de/project/show/Devel:Galaxy:Manager:4.1:SLE12-SUSE-Manager-Tools) and [SLE15 client tools](https://build.suse.de/project/show/Devel:Galaxy:Manager:4.1:SLE15-SUSE-Manager-Tools).

**Comment by juliogonzalez** [comment]
> We are not actually supporting grafana itself for other OSs (RES or Ubuntu), only for SLE12 and SLE15. Therefore - from my perspective - we should just ship Dario's dashboard packages alongside the grafana package with our SLE12 and SLE15 client tools.

Ok, then we need @MalloZup to prepare a submission to the SLE12 codestreams (SUSE:SLE-12:Update, and from there it should get into all SPs), as I understand from his comment that he will be the maintainer.

With that, we can ask maintenance to modify the channel definitions to take the packages from the SLE channels.

**Comment by MalloZup** [comment]
@juliogonzalez Yes for SLE12 we have in our backlog a plan to submit ECO for dashboards

**Comment by juliogonzalez** [comment]
@MalloZup what will be the package?

Ideally we should have that package available at SLE12 before the second week of September. If it's a new package that for now only SUSE Manager is going to use, you don't really need an ECO if you just want to submit to SUSE:SLE-12:Update, AFAIK

**Comment by juliogonzalez** [comment]
Also we'll need this updated grafana-formula package submitted to the Devel:Galaxy:Manager:4.1, Devel:Galaxy:Manager:Head and systemsmanagement:Uyuni:Master. I will create a card for you @renner

File: grafana-formula/grafana-formula.spec
Comment (witekest) [inline grafana-formula/grafana-formula.spec:22]: nit: as we add new dashboards we should bump the minor version number

File: grafana-formula/metadata/form.yml
Comment (witekest) [inline grafana-formula/metadata/form.yml:94]: As we add more and more dashboards the list will get pretty long. Creating sub-groups per product (Kubernetes and SAP for now) would help structuring it a little.

Comment (MalloZup) [reply grafana-formula/metadata/form.yml:94]: @witekest  unfortunately our 3 dashboards cannot be grouped because we have 3 different components which need to be installed as standalone.  (HA, hana and netweaver dashboard).

For example, a customer  might want just to install  HA  (highly availability) without the SAP components.

Also for HANA, this doesn't imply that he wants netweaver.

So having this separated is the only solution for this scenarios

**Comment by MalloZup** [comment]
thx for all reviews.

resuming the follow-up todos:

- [ ] fix review comment and test formula with RPM (@MalloZup )
- [ ] submit the packages to SLE12 as discussed with @juliogonzalez 

I guess after  this we should be ok to merge the PR

**Comment by witekest** [comment]
> I guess after this we should be ok to merge the PR, or I'm missing something?

If you want to split the work into smaller steps I'd suggest creating a feature branch and continuing the work there.
I'm against merging it into _master_ in current shape as the [syntax errors](https://github.com/SUSE/salt-formulas/pull/67#pullrequestreview-466024169) break the installation of existing dashboards.

**Comment by MalloZup** [comment]
@witekest no worries, if you check my first point is about what you suggest. see https://github.com/SUSE/salt-formulas/pull/67#issuecomment-673398111 :sunflower:

**Comment by juliogonzalez** [comment]
> I guess after this we should be ok to merge the PR, or I'm missing something?

There's one extra task, which is for SUSE Manager to update grafana-formula on the SUSE Manager and Uyuni codestreams  as the package is only there and NOT at the SLE/openSUSE codestreams, and I guess we want to keep it that way).

I created an internal card for it.

File: grafana-formula/metadata/form.yml
Comment (witekest) [reply grafana-formula/metadata/form.yml:94]: I thought about visual structuring only. Adding two subgroups would work, e.g.:

```
dashboards:
  ...

  kubernetes:
    $type: group
    $help: Kubernetes dashboards to install.

    add_k8s_dashboard:
    ... 

  SAP:
    $type: group
    $help: SAP dashboards to install.

    add_ha_dashboard:
    ... 
```

**Comment by MalloZup** [comment]
@witekest thx for taking care of this. we need to wait to merge this until grafana dashboards lands to sle15 ..

**Review by witekest — APPROVED** [review]

**Comment by juliogonzalez** [comment]
@MalloZup the packages are on SLE12 and SLE15, and you have two approvals and no blockers. Is there anything else pending?

**Comment by MalloZup** [comment]
@juliogonzalez nope nothing pending for dashboards.

**Comment by MalloZup** [comment]
@witekest  the grafana pkg are merged so we can merge this unless there are other issues

**Comment by juliogonzalez** [comment]
@MalloZup you have conflicts on the changelog, you will need to fix them.

**Comment by MalloZup** [comment]
@juliogonzalez the conflicts where caused from 2 hours ago:
https://github.com/SUSE/salt-formulas/pull/81#event-3839555174
:(

**Comment by juliogonzalez** [comment]
If this is ready to be merged, I suggest you go ahead with it so @witekest can prepare the SRs :-)

**Pull request merged**
